### PR TITLE
ON HOLD: REST API test framework

### PR DESCRIPTION
**Pull request by DanGraur**
Added a test mini-framework for IPv8's REST API in the ipv8.test.REST.rtest package.

### CODE REVIEW

**Comment by qstokkink** [comment]
Ok to test

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

File: ipv8/REST/attestation_endpoint.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ def render_GET(self, request):
         type=peers -> [mid_b64]
         type=attributes&mid=mid_b64 -> [(attribute_name, attribute_hash)]
         """
+
         if not request.args or 'type' not in request.args:
```
Comment (qstokkink) [inline ipv8/REST/attestation_endpoint.py]: Please remove the added blank lines in this file.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+from ipv8.test.util import twisted_wrapper
+
+
+class SingleServerSetup(unittest.TestCase):
+    """AndroidTestPeer
+    Test class which defines an environment with one well-known peer. This should be extended by other subclasses,
+    which implement specific test cases.
+    """
+
+    def initialize(self, **kwargs):
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: Please make the arguments specific with optional values.
For example:
`port=8086, interface='127.0.0.1', configuration=None` etc.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+        :param dict_param: the required parameters by the GET request generator for the peers request type
+        :return: a list of currently known peers in the network
+        """
+
+        peer_list = yield self._get_style_requests.make_peers(dict_param)
+
+        # Keep iterating until peer_list is non-empty
+        while peer_list == "[]":
+            # Wait for 4 seconds before trying again
+            time.sleep(4)
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: This does not do what you think it does while running on the Twisted thread. By sleeping you actually block the entirety of Python and almost nothing else will happen. Use `yield deferLater(reactor, time, lambda: None)` instead.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+
+        :param dict_param: the required parameters by the GET request generator for the outstanding request type
+        :return: a list of outstanding attestation requests
+        """
+        outstanding_requests = yield self._get_style_requests.make_outstanding(dict_param)
+
+        # Keep iterating until peer_list is non-empty
+        while outstanding_requests == "[]":
+            # Wait for 4 seconds before trying again
+            time.sleep(4)
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: Same here.

File: ipv8/test/REST/rtest/test_rest_api_peer.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+class InteractiveTestPeer(TestPeer, threading.Thread):
+    """
+    This class models the basic behavior of simple peer instances which are used for interaction. Subclasses should
+    implement the actual main logic of the peer in the run() method (from Thread).
+    """
+
+    def __init__(self,
+                 path,
+                 port,
+                 **kwargs):
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/test_rest_api_peer.py]: Same here.

File: ipv8/test/REST/rtest/test_rest_api_peer.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                    'peers': 20,
+                    'init': {
+                        'timeout': 60.0
+                    }
+                }]
+
+        self._create_working_directory(self._path)
+        self._logger.info("Created working directory.")
+        os.chdir(self._path)
+
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/test_rest_api_peer.py]: Please overwrite the AttestationCommunity and IdentityCommunity `master_peer` so you do not interfere with the live network (this testsuite was firing requests to the live server :) ).

**Comment by DanGraur** [comment]
Resolved the issues in the Code Review. I think it's ready for another review, since I added some additional code.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+        peer_list = set(ast.literal_eval(peer_list))
+
+        # Keep iterating until peer_list is non-empty
+        while not peer_list - excluded_peer_mids:
+            # Wait for 4 seconds before trying again
+            yield deferLater(reactor, 4, lambda: None)
+
+            # Forward and wait for the response
+            peer_list = yield self._get_style_requests.make_peers(dict_param)
+            peer_list = set(ast.literal_eval(peer_list))
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: Whoa, this allows for arbitrary code execution. This should definitely use `json` parsing.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+
+        :param dict_param: the required parameters by the GET request generator for the outstanding request type
+        :return: a list of outstanding attestation requests
+        """
+        outstanding_requests = yield self._get_style_requests.make_outstanding(dict_param)
+
+        # Keep iterating until peer_list is non-empty
+        while outstanding_requests == "[]":
+            # Wait for 4 seconds before trying again
+            yield deferLater(reactor, 4, lambda: None)
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: As this all happens on the same host anyway, you could just use a `0.1` second delay instead.

File: ipv8/test/REST/rtest/rest_peer_communication.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        :param arguments: the arguments to be attached to the request
+        :param on_complete_callback:
+        :return: a Deferred object for the response of this request
+        """
+        # If no arguments are sent, then assign default empty arguments
+        if arguments is None:
+            arguments = {}
+
+        request_url = url + '?' + '&'.join("%s=%s" % (k, v) for k, v in arguments.iteritems())
+        print "\t[HTTP-%s] %s" % (request_type, request_url)
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/rest_peer_communication.py]: Please use `logging.info()` or `logging.debug()` for this in the tests.

File: ipv8/test/REST/rtest/test_rest_api_peer.py [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+    'TunnelCommunity': ECCrypto().generate_key(u'high')
+}
+
+
+class TestPeer(object):
+    """
+    Class for the purpose of testing the REST API
+    """
+
+    master_peer_att = Peer(("3052301006072a8648ce3d020106052b8104001a033e000400fde127289850e6e550d28a083"
```
Comment (qstokkink) [inline ipv8/test/REST/rtest/test_rest_api_peer.py]: These appear to be no longer used.

File: ipv8/test/REST/rtest/peer_interaction_test_env.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+        :param interface: IP or alias of the peer. Defaults to '127.0.0.1'.
+        :param configuration: IPv8 configuration object. Defaults to None.
+        :param get_style_requests: GET style request generator. Defaults to None.
+        :param post_style_requests: POST style request generator. Defaults to None.
+        :return:
+        """
+        assert get_style_requests is None or isinstance(get_style_requests, GetStyleRequests), \
+            "The get_style_requests parameter must be a subclass of GetStyleRequests"
+        assert post_style_requests is None or isinstance(post_style_requests, PostStyleRequests), \
+            "The get_style_requests parameter must be a subclass of GetStyleRequests"
```
Comment (xoriole) [inline ipv8/test/REST/rtest/peer_interaction_test_env.py]: please update this message

**Comment by qstokkink** [comment]
Don't forget to clean your commit history.

**Comment by DanGraur** [comment]
I've moved these changes to a new PR #233. I'll delete this PR soon, after I finish work on that one, since there are some unit tests which are currently failing.

**Pull request closed without merging**